### ISSUE TITLE
[Tracking Issue] Hardware enablement policies in nixpkgs

### ISSUE THREAD

**Issue by RaitoBezarius** [NixOS/nixpkgs#301541]
## Issue description

We have https://github.com/NixOS/nixos-hardware/ which is a repo where everyone dumps pieces of configurations they find to make their computer do better in their view.

In the past, that repo was merged _inside_ nixpkgs, then reverted, in the context of https://github.com/NixOS/nixpkgs/pull/91160.

Furthermore, `nixos-generate-config` is not always aware of potential hardware profiles to apply, see https://github.com/NixOS/nixos-hardware/issues/49 for ideas, which impedes the discoverability sometimes.

We discussed on Matrix at https://matrix.to/#/!kjdutkOsheZdjqYmqp:nixos.org/$2Tu2fydLwl5XTeuHQxOrqGa0TAn0F6ft4e6fdylQaAU?via=lossy.network&via=matrix.org&via=nixos.dev "what if we merged the repo again" and we came to the conclusion very quickly this was not an option.

Here's a summary of the discussion:

- Kernel maintainers in nixpkgs are not super comfortable with forked kernels _inside_ of nixpkgs: @alyssais and I prefer a hard rule "no forked kernel" even if it had permitted hardware enablement, @K900 is fine with *explicitly temporary*.
- A code owner of `hardware/`, even multiples to ensure a high quality level, on par with Jovian or Asahi
- For something to be present in `hardware/`, the intention of their vendor/author is that it should be upstreamed as soon as possible
- Explicit traces of good maintenance from the midstream (Asahi/Collabora/Jovian) must be apparent
- We also mentioned that CI in nixos-hardware is lacking and a better separation of packages/modules would help

#### Raito's proposal

My proposal would be to formalize better a policy in Nixpkgs about the hardware enablement here, take ownership of that repo and start to look if some stuff makes sense to move over from nixos-hardware to here.

In parallel, we can help the nixos-hardware project to get CI and help with the structure.

Finally, I'd like to try to close the "auto generate" profile issue with `nixos-generate-config`, during the conversation: https://git.numtide.com/numtide/nixos-generate-facts was mentioned.

cc @samueldr @zimbatm

**Comment by zimbatm** [comment]
Reading trough the lines, the problem you want to solve is to improve the quality of nixos-generate-config. Is that correct?

**Comment by RaitoBezarius** [comment]
> Reading trough the lines, the problem you want to solve is to improve the quality of nixos-generate-config. Is that correct?

Not really, this is about hardware enablement policies in general, `nixos-generate-config` improvements would be a step / sub-issue of this one, yes.

At least, having `ngc` able to add more known profiles/tweaks based on nixpkgs and maybe... nixos-hardware with a warning this comes from a community source with less quality standards, etc.

**Comment by nyabinary** [comment]
I definitely agree with the sentiment of moving some things from nixos-hardware to nixpkgs, maybe a first good candidate would be the Framework module? It seems pretty well written, also how would the structure of this work? Could ``nixos-generate-config`` automatically determine which computer the user has and install that specific profile for their machine?

**Comment by sdht0** [comment]
I personally didn't want to enable everything suggested in [nixos-hardware/](https://github.com/NixOS/nixos-hardware/) automatically. Instead I went through all the options manually and copied the ones I wanted.

An idea I wanted to explore was to automatically generate the merged configuration from all files for the particular hardware I'm interested in so I didn't have to follow all the includes. 

Maybe we can apply that here as the first step: Automatically deploy a website based on the repo where all suggested configuration for each hardware profile will be available to the users with search.  `nixos-generate-config` can then have a comment at the end linking to this website saying: "search for your hardware here for more relevant config options".

**Comment by nyabinary** [comment]
> I personally didn't want to enable everything suggested in [nixos-hardware/](https://github.com/NixOS/nixos-hardware/) automatically. Instead I went through all the options manually and copied the ones I wanted.
> 
> An idea I wanted to explore was to automatically generate the merged configuration from all files for the particular hardware I'm interested in so I didn't have to follow all the includes.
> 
> Maybe we can apply that here as the first step: Automatically deploy a website based on the repo where all suggested configuration for each hardware profile will be available to the users with search. `nixos-generate-config` can then have a comment at the end linking to this website saying: "search for your hardware here for more relevant config options".

I mean that can be fixed, nixos-hardware can be the "staging" repo before merging stuff into nixpkgs, that way we can make sure everything is unopinionated, and then optimally we would have those more opinionated things as an opt-in somehow.

**Comment by RossComputerGuy** [comment]
> * Kernel maintainers in nixpkgs are not super comfortable with forked kernels _inside_ of nixpkgs: @alyssais, @K900 and I prefer a hard rule "no forked kernel" even if it had permitted hardware enablement.

> * A code owner of `hardware/`, even multiples to ensure a high quality level, on par with Jovian or Asahi

Would Asahi support be included? Imo, it would be great and I would even be willing to maintain that inside of Nixpkgs. However, in Asahi's current state, their kernel is a hard requirement. Would there be exceptions to the forked kernel rule?

**Comment by K900** [comment]
Asahi would be included under nixos-hardware, but not as an "official package".
